MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc

```diff
@@ -47,62 +51,71 @@ volumeHandle::volumeHandle(const cms::DDFilteredView &fv, bool expand2Pi, bool d
   switch (theShape) {
     case DDSolidShape::ddbox: {
       DDBox box(solid.solid());
-      // DD4hep returns units in cm, no conversion needed.
-      double halfX = box.x();
-      double halfY = box.y();
-      double halfZ = box.z();
+      // Convert from DD4hep units to cm
+      double halfX = box.x() / dd4hep::cm;
+      double halfY = box.y() / dd4hep::cm;
+      double halfZ = box.z() / dd4hep::cm;
       buildBox(halfX, halfY, halfZ);
     } break;
     case DDSolidShape::ddtrap: {
       DDTrap trap(solid.solid());
-      double x1 = trap.bottomLow1();
-      double x2 = trap.topLow1();
-      double x3 = trap.bottomLow2();
-      double x4 = trap.topLow2();
-      double y1 = trap.high1();
-      double y2 = trap.high2();
+      double x1 = trap.bottomLow1() / dd4hep::cm;
+      double x2 = trap.topLow1() / dd4hep::cm;
+      double x3 = trap.bottomLow2() / dd4hep::cm;
+      double x4 = trap.topLow2() / dd4hep::cm;
+      double y1 = trap.high1() / dd4hep::cm;
+      double y2 = trap.high2() / dd4hep::cm;
       double theta = trap.theta();
       double phi = trap.phi();
-      double halfZ = trap.dZ();
+      double halfZ = trap.dZ() / dd4hep::cm;
       double alpha1 = trap.alpha1();
       double alpha2 = trap.alpha2();
       buildTrap(x1, x2, x3, x4, y1, y2, theta, phi, halfZ, alpha1, alpha2);
     } break;
 
     case DDSolidShape::ddcons: {
       DDCons cons(solid.solid());
-      double zhalf = cons.dZ();
-      double rInMinusZ = cons.rMin1();
-      double rOutMinusZ = cons.rMax1();
-      double rInPlusZ = cons.rMin2();
-      double rOutPlusZ = cons.rMax2();
+      double zhalf = cons.dZ() / dd4hep::cm;
+      double rInMinusZ = cons.rMin1() / dd4hep::cm;
+      double rOutMinusZ = cons.rMax1() / dd4hep::cm;
+      double rInPlusZ = cons.rMin2() / dd4hep::cm;
+      double rOutPlusZ = cons.rMax2() / dd4hep::cm;
       double startPhi = cons.startPhi();
       double deltaPhi = reco::deltaPhi(cons.endPhi(), startPhi);
       buildCons(zhalf, rInMinusZ, rOutMinusZ, rInPlusZ, rOutPlusZ, startPhi, deltaPhi);
     } break;
     case DDSolidShape::ddtubs: {
       DDTubs tubs(solid.solid());
-      double zhalf = tubs.dZ();
-      double rIn = tubs.rMin();
-      double rOut = tubs.rMax();
+      double zhalf = tubs.dZ() / dd4hep::cm;
+      double rIn = tubs.rMin() / dd4hep::cm;
+      double rOut = tubs.rMax() / dd4hep::cm;
       double startPhi = tubs.startPhi();
       double deltaPhi = tubs.endPhi() - startPhi;
       buildTubs(zhalf, rIn, rOut, startPhi, deltaPhi);
     } break;
     case DDSolidShape::ddpseudotrap: {
       vector<double> d = solid.parameters();
+      if (d.size() >= 7)
+        LogTrace("MagGeoBuilder") << " Pseudo trap params raw =  " << d[0] << ", " << d[1] << ", " << d[2] << ", "
+                                  << d[3] << ", " << d[4] << ", " << d[5] << ", " << d[6];
+
+      transform(d.begin(), --(d.end()), d.begin(), [](double val) { return val / dd4hep::cm; });
```

Comment:
is it intented to transform all inputs **except** the last one in the list? Given the debugging comments around this line, I suppose the last parameter d[6] has some special meaning, which I was not able to dig up immediately. Perhaps a code comment could clarify things for future readers?

Reply:
yes, indeed d[6] is to be interpreted as a bool, cf:
https://github.com/cms-sw/cmssw/blob/134ed3345e18f52b5c8e50e3217d326b760b6402/MagneticField/GeomBuilder/src/buildPseudoTrap.icc#L16
That should be clear from the signature of that function, but I agree a comment here would not harm.

Reply:
Yes, please put a comment here, to make a bit clearer that the transform relies on the special signature of the `buildPseudoTrap` function.

Please also change the following to be even more explicit

```C++
buildPseudoTrap(d[0], d[1], d[2], d[3], d[4], d[5], static_cast<bool>(d[6]));
```

I remark here that doing bool->double->bool seems a bit odd in general, but I don't know enough about this code to suggest a better alternative right now.

Reply:
Sorry, I should have put a comment here. The pseudo-trapezoid is a special case that is only used for legacy configurations. I will add clarifications.